File: cdm/cpp/io/protobuf/PBEvents.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "io/protobuf/PBEvents.h"
#include "io/protobuf/PBProperties.h"
#include "io/protobuf/PBUtils.h"
#include "bind/cpp/cdm/Events.pb.h"
#include "engine/SEEventManager.h"
#include "utils/FileUtils.h"

const std::string& eEvent_Name(eEvent m)
{
  return cdm::eEvent_Name((cdm::eEvent)m);
}

bool PBEvents::SerializeToString(std::vector<const SEEventChange*>& changes, std::string& output, SerializationFormat m, Logger* logger)
{
  if (changes.empty())
    return false;

  cdm::EventChangeListData data;
  for (const SEEventChange* c : changes)
  {
    cdm::EventChangeData* change = data.add_change();
    change->set_event((cdm::eEvent)c->GetEvent());
    change->set_active(c->GetActive());
    if (c->GetSimTime().IsValid())
      change->set_allocated_simtime(PBProperty::Unload(c->GetSimTime()));
  }
  if (!PBUtils::SerializeToString(data, output, m, logger))
    return false;
  return true;
}

bool PBEvents::SerializeFromString(const std::string& src, std::vector<const SEEventChange*>& changes, SerializationFormat m, Logger* logger)
{
  SEScalarTime time;
  cdm::EventChangeListData data;
  if (!PBUtils::SerializeFromString(src, data, m, logger))
    return false;
  for (int i = 0; i < data.change_size(); i++)
  {
    auto& e = data.change()[i];
    if (e.has_simtime())
      PBProperty::Load(e.simtime(), time);
    else
      time.Invalidate();
    SEEventChange* c = new SEEventChange((eEvent)e.event(), e.active(), &time);
    changes.push_back(c);
  }

  return true;
}

bool PBEvents::SerializeToString(std::vector<const SEActiveEvent*>& active_events, std::string& output, SerializationFormat m, Logger* logger)
{
  if (active_events.empty())
    return false;

  cdm::ActiveEventListData data;
  for (const SEActiveEvent* a : active_events)
  {
    cdm::ActiveEventData* aeData = data.add_activeevent();
    aeData->set_event((cdm::eEvent)a->GetEvent());
    if (a->GetDuration().IsValid())
      aeData->set_allocated_duration(PBProperty::Unload(a->GetDuration()));
  }
  if (!PBUtils::SerializeToString(data, output, m, logger))
    return false;
  return true;
}

bool PBEvents::SerializeFromString(const std::string& src, std::vector<const SEActiveEvent*>& active_events, SerializationFormat m, Logger* logger)
{
  SEScalarTime time;
  cdm::ActiveEventListData data;
  if (!PBUtils::SerializeFromString(src, data, m, logger))
    return false;
  for (int i = 0; i < data.activeevent_size(); i++)
  {
    auto& ae = data.activeevent()[i];
    if (ae.has_duration())
      PBProperty::Load(ae.duration(), time);
    else
      time.Invalidate();
    SEActiveEvent* c = new SEActiveEvent((eEvent)ae.event(), time);
    active_events.push_back(c);
  }

  return true;
}

File: cdm/csharp/io/PBEvents.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

using System.Collections.Generic;
using Google.Protobuf;

namespace Pulse.CDM
{
  public class PBEvents
  {
    public static void Load(Cdm.ActiveEventListData src, List<SEActiveEvent> dst)
    {
      Serialize(src, dst);
    }
    public static void Serialize(Cdm.ActiveEventListData src, List<SEActiveEvent> dst)
    {
      SEScalarTime duration = new SEScalarTime();
      foreach (var active in src.ActiveEvent)
      {
        PBProperty.Load(active.Duration, duration);
        dst.Add(new SEActiveEvent((eEvent)active.Event, duration));
      }
    }
    public static bool SerializeFromString(string src, List<SEActiveEvent> dst)
    {
      try
      {
        Cdm.ActiveEventListData data = JsonParser.Default.Parse<Cdm.ActiveEventListData>(src);
        PBEvents.Load(data, dst);
      }
      catch (Google.Protobuf.InvalidJsonException)
      {
        return false;
      }
      return true;
    }

    public static Cdm.ActiveEventListData Unload(List<SEActiveEvent> src)
    {
      Cdm.ActiveEventListData dst = new Cdm.ActiveEventListData();
      Serialize(src, dst);
      return dst;
    }
    public static void Serialize(List<SEActiveEvent> src, Cdm.ActiveEventListData dst)
    {
      foreach (var active in src)
      {
        Cdm.ActiveEventData ae = new Cdm.ActiveEventData();
        ae.Event = (Cdm.eEvent)active.Event;
        ae.Duration = PBProperty.Unload(active.Duration);
        dst.ActiveEvent.Add(ae);
      }
    }
    public static string SerializeToString(List<SEActiveEvent> src)
    {
      var pb = PBEvents.Unload(src);
      return pb.ToString();
    }

    public static void Load(Cdm.EventChangeListData src, List<SEEventChange> dst)
    {
      Serialize(src, dst);
    }
    public static void Serialize(Cdm.EventChangeListData src, List<SEEventChange> dst)
    {
      SEScalarTime simTime = new SEScalarTime();
      foreach (var change in src.Change)
      {
        PBProperty.Load(change.SimTime, simTime);
        dst.Add(new SEEventChange((eEvent)change.Event, change.Active, simTime));
      }
    }
    public static bool SerializeFromString(string src, List<SEEventChange> dst)
    {
      try
      {
        Cdm.EventChangeListData data = JsonParser.Default.Parse<Cdm.EventChangeListData>(src);
        PBEvents.Load(data, dst);
      }
      catch (Google.Protobuf.InvalidJsonException)
      {
        return false;
      }
      return true;
    }

    public static Cdm.EventChangeListData Unload(List<SEEventChange> src)
    {
      Cdm.EventChangeListData dst = new Cdm.EventChangeListData();
      Serialize(src, dst);
      return dst;
    }
    public static void Serialize(List<SEEventChange> src, Cdm.EventChangeListData dst)
    {
      foreach (var change in src)
      {
        Cdm.EventChangeData c = new Cdm.EventChangeData();
        c.Event = (Cdm.eEvent)change.Event;
        c.Active = change.Active;
        if (change.SimTime.IsValid())
          c.SimTime = PBProperty.Unload(change.SimTime);
        dst.Change.Add(c);
      }
    }
    public static string SerializeToString(List<SEEventChange> src)
    {
      var pb = PBEvents.Unload(src);
      return pb.ToString();
    }
  }
}
